File: python/src/os_vars.py
#!/usr/bin/env python

#############################################################################
# HEADER                                                                    #
#############################################################################
"""
encapsulation to environment variables definition, validation and access
"""

#############################################################################
# IMPORT MODULES                                                            #
#############################################################################
import os
import sys
from .logger import Logger

# supposedly, the only unregistered / declared env vars!
DUMP_CONFIG_ENV_VAR = "__DUMP_CONFIG"
PRINT_CONFIG_USAGE_ENV_VAR = "__CONFIG_USAGE"

#############################################################################
# IMPLEMENTATION                                                            #
#############################################################################


class OSVars:
    """
    The OSVars class serving application with clarifying their dependant and expected os/env variables.
    One may use os.environ to get that but when all goes through this class the benefits are:

    1. Encapsulation - no more env var per sporadic demand in sporadic py file
    2. Usage - user can get all of the dependant/expected env vars from the service it is interacting with along with
       their description and default values.
    3. Type validation - env var can be defined to be of type X (ex int) so when it is provided with an unexpected
       value type - the process exists (to prevent unexpected behavior). For unknown types register var as str.
    4. Dump - running service can dump all of its vars and their actual values served to the application (for better analysis)


    Class Public Interface:

      register
      register_mandatory
      initialize
      get
      usage
      dump
    """

    __instance = None

    @staticmethod
    def instance():
        """ Static access method. """
        if OSVars.__instance is None:
            # the only place the Secrets constructor is being called.
            OSVars()
        return OSVars.__instance

    def __init__(self):
        if OSVars.__instance is None:
            OSVars.__instance = self
        else:
            # should anyone call Secrets directly, make a bold statement about it.
            raise Exception(
                """OSVars object already initialized - you cannot create another instance!
                (hint: use OSVars.instance()"""
            )
        self.__initialized = False
        # The registered vars mem db
        self.__vars = {}
        # by default, when critical requirement is invalid then sys.exit except when dumping
        self.__can_exit = (
            PRINT_CONFIG_USAGE_ENV_VAR not in os.environ
            and DUMP_CONFIG_ENV_VAR not in os.environ
        )
        self.__should_print_usage = PRINT_CONFIG_USAGE_ENV_VAR in os.environ
        self.__should_dump_config = DUMP_CONFIG_ENV_VAR in os.environ

    @staticmethod
    def initialize():
        OSVars.instance().__initialize()

    def __initialize(self):
        self.__validate_and_set()
        self.__initialized = True

    def __critical_fault(self, message):
        Logger.error(f"ENV VAR ERROR: {message}")
        if self.__can_exit is True:
            sys.exit(1)

    @staticmethod
    def register(var_key, var_description, var_type=str, default_value=None):
        return OSVars.instance().__register(
            var_key, var_description, var_type, False, default_value
        )

    @staticmethod
    def register_mandatory(var_key, var_description, var_type=str, default_value=None):
        return OSVars.instance().__register(
            var_key, var_description, var_type, True, default_value
        )

    def __validate_and_set(self):
        """
        validates all defined env vars to:
        1. exist in os.environ if var is mandatory
        2. have provided value that adheres to defined type
        Then it sets mem db (__vars) value to the actual (default or os.environ provided)
        """
        for (var_key, var_obj) in self.__vars.items():
            # default case if the other conditions dont apply
            value = var_obj["default"]

            if var_key not in os.environ:
                if var_obj["is_mandatory"] is True:
                    self.__critical_fault(
                        f"Missing mandatory os env var {var_key} ({var_obj['description']})"
                    )
            else:
                value = os.environ[var_key]

            # type checking and casting
            if var_obj["var_type"] != str and value is not None:
                try:
                    value = var_obj["var_type"](value)
                except Exception as ex:
                    self.__critical_fault(
                        f"provided value for {var_key} is expected to be {var_obj['var_type']} but its not (actual: {type(value)})\nDetailed exception: {type(ex)}: {ex}"
                    )

            var_obj["value"] = value

    def __register(
        self,
        var_key,
        var_description,
        var_type=str,
        is_mandatory=False,
        default_value=None,
    ):
        # sanity checks...
        if var_key in self.__vars:
            print(f"Warning: os var {var_key} is already registered!")
            return

        if default_value is not None and is_mandatory is True:
            self.__critical_fault(
                f"defining var {var_key} as mandatory with default value doesn't make sense!"
            )

        self.__vars[var_key] = {
            "description": var_description,
            "var_type": var_type,
            # "value": value, # responsibility of __validate_and_set
            "is_mandatory": is_mandatory,
            "default": default_value,
        }

    @staticmethod
    def get(var_key):
        return OSVars.instance().__get(var_key)

    def __get(self, var_key):
        if self.__should_print_usage is True:
            self.usage()
            sys.exit(1)
        elif self.__should_dump_config is True:
            self.dump()
            # this is a one time endeavor
            self.__should_dump_config = False

        if self.__initialized is False:
            raise Exception("OSVars has not been initialized. call OSVars.initialize()")

        if var_key not in self.__vars:
            raise Exception(
                f"{var_key} unknown. Please specify variable attributes using the register method in the process initialization(!)"
            )

        return self.__vars[var_key]["value"]

    def usage(self):
        if self.__initialized is False:
            raise Exception("OSVars has not been initialized. call OSVars.initialize()")

        mandatory_sign = {0: "", 1: "* "}

        print("\n\nEnvironment variables usage:\n")
        for (var_key, var_obj) in self.__vars.items():
            default = (
                f". (Default: {var_obj['default']})"
                if var_obj["default"] is not None
                else ""
            )
            print(
                f"{mandatory_sign[var_obj['is_mandatory']]}{var_key} ({var_obj['var_type'].__name__}): {var_obj['description']}{default}"
            )
        print(
            f"""
        * - mandatory vars
        {PRINT_CONFIG_USAGE_ENV_VAR}: set to any to produce this usage print
        {DUMP_CONFIG_ENV_VAR}: set to any to dump actual env vars values"""
        )

    def dump(self):
        if self.__initialized is False:
            raise Exception("OSVars has not been initialized. call OSVars.initialize()")

        print("\n\nEnvironment variables mem dump:\n")
        for (var_key, var_obj) in self.__vars.items():
            default = (
                f". (Default: {var_obj['default']})"
                if var_obj["default"] is not None
                else ""
            )
            print(
                f"{var_key}: {var_obj['value']} \n\t{var_obj['description']}{default}"
            )

File: ruby/src/os_vars.rb
require 'singleton'
require_relative 'utils/type_utils.rb'

module TwistConf
  #############################################################################
  # GLOBALS and CONSTANTS                                                     #
  #############################################################################

  # supposedly, the only unregistered / declared env vars!
  DUMP_CONFIG_ENV_VAR = '__DUMP_CONFIG'
  PRINT_CONFIG_USAGE_ENV_VAR = '__CONFIG_USAGE'

  #############################################################################
  # IMPLEMENTATION                                                            #
  #############################################################################

  class OSVars
    include Singleton

    def initialize
      @__initialized = false

      # The registered vars mem db
      @__vars = {}
      # by default, when critical requirement is invalid then sys.exit except when dumping
      @__can_exit = (
          ENV[PRINT_CONFIG_USAGE_ENV_VAR].nil? &&
          ENV[DUMP_CONFIG_ENV_VAR].nil?)

      @__should_print_usage = !ENV[PRINT_CONFIG_USAGE_ENV_VAR].nil?
      @__should_dump_config = !ENV[DUMP_CONFIG_ENV_VAR].nil?
    end

    # validates all defined env vars to:
    # 1. exist in ENV if var is mandatory
    # 2. have provided value that adheres to defined type
    # Then it sets mem db (__vars) value to the actual (default or ENV provided)
    #
    # @return [void or throws]
    def __validate_and_set
      @__vars.each do |var_key, var_obj|
        # default case if the other conditions dont apply
        value = var_obj['default']

        if ENV[var_key].nil?
          if var_obj['is_mandatory'] == true
            __critical_fault("Missing mandatory os env var #{var_key} (#{var_obj['description']})")
          end
        else
          value = ENV[var_key]
        end

        # type checking and casting
        if var_obj['var_type'].name != 'String' && !value.nil?
          begin
            value = dynamic_cast(var_obj['var_type'], value)
          rescue StandardError => e
            __critical_fault("provided value for #{var_key} is expected to be #{var_obj['var_type']} but its not \n" \
              "Detailed exception:  #{e}")
          end
        end

        # finally
        var_obj['value'] = value
      end
    end

    # method that signs the registration - must be called post process initialization!
    #
    # @return [void or throws]
    def init
      __validate_and_set
      @__initialized = true
    end

    # helper method to std output error and exit (hence critical)
    #
    # @param message [String] - error message
    # @return - exits...
    def __critical_fault(message)
      puts "ENV VAR ERROR: #{message}"
      exit(1) unless @__can_exit == false
    end

    # helper static method for easier access (OSVars.register)
    def self.register(var_key, var_description, var_type = String, default_value = nil)
      OSVars.instance.__register(var_key, var_description, var_type, false, default_value)
    end

    # helper static method for easier access (OSVars.register_mandatory)
    def self.register_mandatory(var_key, var_description, var_type = String, default_value = nil)
      OSVars.instance.__register(var_key, var_description, var_type, true, default_value)
    end

    # os var registration / requirement method
    #
    # @param var_key [String] - name of env var
    # @param var_description [String] - elaborated purpose of this env within ones process
    # @param var_type [Class] - expected type for ENV value (will be validated converted!)
    # @param is_mandatory [bool] - if true, process will exit if var not found
    # default_value [String] - returned value for optional var that dont exist in ENV
    #
    # @return [void or throws]
    def __register(var_key, var_description, var_type = String, is_mandatory = false, default_value = nil)
      # sanity checks...
      unless @__vars[var_key].nil?
        puts "os var #{var_key} is already registered!"
      end

      # another sanity check...
      if !default_value.nil? && is_mandatory == true
        __critical_fault("defining var #{var_key} as mandatory with default value doesn't make sense!")
      end

      # the actual registration
      @__vars[var_key] = {
        'description' => var_description,
        'var_type' => var_type,
        # "value": value, # responsibility of __validate_and_set
        'is_mandatory' => is_mandatory,
        'default' => default_value
      }
    end

    # helper static method for easier access (OSVars.get)
    def self.get(var_key)
      OSVars.instance.__get(var_key)
    end

    # main os env var value accessor
    #
    # @param var_key [String] - name of os var name
    # @return [Class or throws or exit]
    #
    # side effect - only upon first time call - print usage or dump actual var + value list
    def __get(var_key)
      if @__should_print_usage == true
        usage
        exit(1)
      end
      if @__should_dump_config == true
        dump
        # this is a one time endeavor
        @__should_dump_config = false
      end

      # the below is not imperative when usage and dump are requested.
      if @__initialized == false
        raise 'OSVars has not been initialized. call OSVars.instance.init'
      end

      # all vars must be known in advance
      if @__vars[var_key].nil?
        raise "#{var_key} unknown. Please specify variable attributes using " \
        'the register method in the process initialization(!)'
      end

      # epilogue
      @__vars[var_key]['value']
    end

    # helper function to printing all process registered env vars, thier types, description and default value
    #
    # @return [void]
    def usage
      unless @__initialized
        raise 'OSVars has not been initialized. call OSVars.initialize()'
      end

      mandatory_sign = { false => '', true => '* ' }

      puts "\n\nEnvironment variables usage:\n\n"
      @__vars.each do |var_key, var_obj|
        default =
          if !var_obj['default'].nil?
            ". (Default: #{var_obj['default']})"
          else
            ''
          end
        puts "#{mandatory_sign[var_obj['is_mandatory']]}#{var_key} " \
        "(#{var_obj['var_type'].name}): #{var_obj['description']}#{default}"
      end
      # epilogue
      puts "\n* - mandatory vars\n" \
      "#{PRINT_CONFIG_USAGE_ENV_VAR}: set to any to produce this usage print\n" \
      "#{DUMP_CONFIG_ENV_VAR}: set to any to dump actual env vars values\n\n"
    end

    # helper function that dumps all the registered os vars and thier actual used values
    #
    # @return [void]
    def dump
      unless @__initialized
        raise 'OSVars has not been initialized. call OSVars.initialize()'
      end

      puts "\n\nEnvironment variables mem dump:\n\n"
      @__vars.each do |var_key, var_obj|
        default =
          if !var_obj['default'].nil?
            ". (Default: #{var_obj['default']})"
          else
            ''
          end
        puts "#{var_key}: #{var_obj['value']} \n\t#{var_obj['description']}#{default}"
      end
    end
  end
end

# see kitchen_sink.rb for better usage examples.
#
# OSVars.register("OREN", "my name", String, "kuku")
# OSVars.register_mandatory("SEA", "my fname", String)
# OSVars.instance.init
# puts "oren is #{OSVars.get("OREN")}"
